add crd process proxy to support some crd based k8s application like spark operator (#991)
jupyter/enterprise_gateway: docs/source/system-architecture.md

within Docker configuration.  Note: because these kernels will always run local to the corresponding Enterprise Gateway instance, these process proxies are of limited use.
- `ConductorClusterProcessProxy` - is responsible for the discovery and management of kernels hosted
within an IBM Spectrum Conductor cluster.
- `SparkOperatorProcessProxy` - is responsible for the descovery and management of kernels hosted 
within a Kubernetes cluster but created as a SparkApplication instead of a Pod. The SparkApplication is a Kubernetes custom resource 
defined inside the project [spark-on-k8s-operator](https://github.com/GoogleCloudPlatform/spark-on-k8s-operator), which 
makes all kinds of spark on k8s components better organized and easy to configure.

> Before you run this kernel, please ensure that spark operator is installed under a certain namespace of your Kubernetes

Reviewer: I'm not sure I understand this statement: _ensure that spark operator is installed under a certain namespace_.

Does the `SparkOperatorProcessProxy` need to know _that_ namespace or are you saying that the spark operator just needs to be deployed with a known namespace?  How is this different from today's kernel pods?

Author: yep, the spark operator includes not only crd(SparkApplication), but also an operator pod which intend to inspect the running status of all running application as well as hook some action when events(start, end, error) happened to any application.

the pod must be deployed under a certain namespace first for start a spark application, so you understanding is correct.

See [Enabling IBM Spectrum Conductor Support](kernel-conductor.html#enabling-ibm-spectrum-conductor-support) for details.

###### CustomResourceProcessProxy

> Enterprise Gateway also provides a implementation of a process proxy derived from KubernetesProcessProxy 

Reviewer: Can we inclose all of the ProcessProxy class references in back-ticks: `KubernetesProcessProxy`, `CustomResourceProcessProxy`.

called CustomResourceProcessProxy. 

> Instead of creating kernel based on Kubernetes pod, CustomResourceProcessProxy
> manages kernel by implementation of custom resource definition(CRD), e.g. SparkApplication is a CRD which include
> plenty component of a Spark on Kubernetes application.

Reviewer: ```suggestion
Instead of creating kernels based on a Kubernetes pod, `CustomResourceProcessProxy`
manages kernels via a custom resource definition (CRD). For example,  `SparkApplication` is a CRD that includes
many components of a Spark-on-Kubernetes application.
```

Author: ok, updated

> If you want to include a new CRD, just wrt [spark_operator.py](https://github.com/jupyter/enterprise_gateway/blob/master/etc/kernel-launchers/kubernetes/scripts/launch_custom_resource.py)

Reviewer: I don't understand the _just wrt_ portion of this statement.  _Just with respect to `spark_operator.py` and `spark_python_operator` kernelspec. _ doesn't really parse for me.  What are you trying to say about this?

Author: my fault

here, i should write down much more detail information about how to quickly extend the crd process proxy

> and [spark_python_operator](https://github.com/abzymeinsjtu/enterprise_gateway/blob/feat_crd_processproxy_support/etc/kernelspecs/spark_python_operator) kernelspec.

Reviewer: This link is pointing to your fork.
